File: rofunc/planning/src/robotics-codes-from-scratch-master/python/LQT_CP.py
'''
    Linear Quadratic tracker with control primitives applied on a via-point example

    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
    Written by Jeremy Maceiras <jeremy.maceiras@idiap.ch>,
    Sylvain Calinon <https://calinon.ch>

    This file is part of RCFS.

    RCFS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 3 as
    published by the Free Software Foundation.

    RCFS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with RCFS. If not, see <http://www.gnu.org/licenses/>.
'''

import numpy as np
import matplotlib.pyplot as plt
from scipy import special
from math import factorial

# Building piecewise constant basis functions
def build_phi_piecewise(nb_data, nb_fct):
    phi = np.kron( np.identity(nb_fct) , np.ones((int(np.ceil(nb_data/nb_fct)),1)) )
    return phi[:nb_data]

# Building radial basis functions (RBFs)
def build_phi_rbf(nb_data, nb_fct):
    t = np.linspace(0,1,nb_data).reshape((-1,1))
    tMu = np.linspace( t[0] , t[-1] , nb_fct )
    phi = np.exp( -1e2 * (t.T - tMu)**2 )
    return phi.T

# Building Bernstein basis functions
def build_phi_bernstein(nb_data, nb_fct):
    t = np.linspace(0,1,nb_data)
    phi = np.zeros((nb_data,nb_fct))
    for i in range(nb_fct):
        phi[:,i] = factorial(nb_fct-1) / (factorial(i) * factorial(nb_fct-1-i)) * (1-t)**(nb_fct-1-i) * t**i
    return phi

# Building Fourier basis functions
def build_phi_fourier(nb_data, nb_fct):

    t = np.linspace(0,1,nb_data).reshape((-1,1))

    # Alternative computation for real and even signal
    k = np.arange(0,nb_fct).reshape((-1,1))
    phi = np.cos( t.T * k * 2 * np.pi )
    return phi.T

# General param parameters
# ===============================

param = {
    "nbData" : 100, # Number of data points
    "nbPoints" : 3, # Number of viapoints
    "nbVarPos" : 2, # Dimension of position data
    "nbDeriv" : 2,  # Number of static and dynamic features (2 -> [x,dx])
    "dt" : 1e-2, # Time step duration
    "nbFct" : 3, # Number of basis function
    "basisName" : "FOURIER", # can be PIECEWEISE, RBF, BERNSTEIN, FOURIER
    "rfactor" : 1e-8 # Control cost
}

nb_var = param["nbVarPos"] * param["nbDeriv"] # Dimension of state vector
R = np.identity( (param["nbData"]-1) * param["nbVarPos"]  ) * param["rfactor"]  # Control cost matrix

tl = np.linspace(0,param["nbData"],param["nbPoints"]+1)
tl = np.rint(tl[1:]).astype(np.int64)-1 
idx_slices = [ slice( i,i+nb_var,1) for i in (tl* nb_var) ]
param["muQ"] = np.zeros(( nb_var * param["nbData"] , 1 ))
Q = np.zeros(( nb_var * param["nbData"] , nb_var * param["nbData"] ))   # Task precision

for slice_t in idx_slices:

    x_t = np.zeros(( nb_var,1 ))
    x_t[:param["nbVarPos"]] = np.random.rand(param["nbVarPos"],1)*5
    
    param["muQ"][slice_t] = x_t
    Q[slice_t,slice_t] = np.diag(  np.hstack(( np.ones(param["nbVarPos"]) , np.zeros( nb_var - param["nbVarPos"])  )) )

# Dynamical System settings (discrete)
# =====================================

A = np.identity( nb_var )
if param["nbDeriv"]==2:
    A[:param["nbVarPos"],-param["nbVarPos"]:] = np.identity(param["nbVarPos"]) * param["dt"]

B = np.zeros(( nb_var , param["nbVarPos"] ))
derivatives = [ param["dt"],param["dt"]**2 /2 ][:param["nbDeriv"]]
for i in range(param["nbDeriv"]):
    B[i*param["nbVarPos"]:(i+1)*param["nbVarPos"]] = np.identity(param["nbVarPos"]) * derivatives[::-1][i]

# Build Sx and Su transfer matrices
Su = np.zeros((nb_var*param["nbData"],param["nbVarPos"] * (param["nbData"]-1))) # It's maybe n-1 not sure
Sx = np.kron(np.ones((param["nbData"],1)),np.eye(nb_var,nb_var))

M = B
for i in range(1,param["nbData"]):
    Sx[i*nb_var:param["nbData"]*nb_var,:] = np.dot(Sx[i*nb_var:param["nbData"]*nb_var,:],A)
    Su[nb_var*i:nb_var*i+M.shape[0],0:M.shape[1]] = M
    M = np.hstack((np.dot(A,M),B)) # [0,nb_state_var-1]

# Building basis funcitons
# =========================

functions = {
    "PIECEWEISE": build_phi_piecewise ,
    "RBF": build_phi_rbf,
    "BERNSTEIN": build_phi_bernstein,
    "FOURIER": build_phi_fourier
}
phi = functions[param["basisName"]](param["nbData"]-1,param["nbFct"])
PSI = np.kron(phi,np.identity(param["nbVarPos"]))

# Batch LQR Reproduction
# =====================================

x0 = np.zeros((nb_var,1))
w_hat = np.linalg.inv(PSI.T @ Su.T @ Q @ Su @ PSI + PSI.T @ R @ PSI) @ PSI.T @ Su.T @ Q @ ( param["muQ"] - Sx @ x0 )
u_hat = PSI @ w_hat
x_hat = (Sx @ x0 + Su @ u_hat).reshape((-1,nb_var))
u_hat = u_hat.reshape((-1,param["nbVarPos"]))

# Plotting
# =========

plt.figure()

plt.title("2D Trajectory")
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

plt.scatter(x_hat[0,0],x_hat[0,1],c='black',s=100)

for slice_t in idx_slices:
    plt.scatter(param["muQ"][slice_t][0],param["muQ"][slice_t][1],c='blue',s=100)

plt.plot( x_hat[:,0] , x_hat[:,1],c='black')

fig,axs = plt.subplots(5,1)

axs[0].plot(x_hat[:,0],c='black')
axs[0].set_ylabel("$x_1$")
axs[0].set_xticks([0,param["nbData"]])
axs[0].set_xticklabels(["0","T"])
for t in tl:
    axs[0].scatter(t,x_hat[t,0],c='blue')

axs[1].plot(x_hat[:,1],c='black')
axs[1].set_ylabel("$x_2$")
axs[1].set_xticks([0,param["nbData"]])
axs[1].set_xticklabels(["0","T"])
for t in tl:
    axs[1].scatter(t,x_hat[t,1],c='blue')

axs[2].plot(u_hat[:,0],c='black')
axs[2].set_ylabel("$u_1$")
axs[2].set_xticks([0,param["nbData"]-1])
axs[2].set_xticklabels(["0","T-1"])

axs[3].plot(u_hat[:,1],c='black')
axs[3].set_ylabel("$u_2$")
axs[3].set_xticks([0,param["nbData"]-1])
axs[3].set_xticklabels(["0","T-1"])

axs[4].set_ylabel("$\phi_k$")
axs[4].set_xticks([0,param["nbData"]-1])
axs[4].set_xticklabels(["0","T-1"])
for i in range(param["nbFct"]):
    axs[4].plot(phi[:,i])
axs[4].set_xlabel("$t$")

plt.show()

File: rofunc/planning/src/robotics-codes-from-scratch-master/matlab/LQT_CP.m
%%    Linear quadratic tracking (LQT) with control primitives applied to a viapoint task (batch formulation)
%%
%%    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
%%    Written by Sylvain Calinon <https://calinon.ch>
%%
%%    This file is part of RCFS.
%%
%%    RCFS is free software: you can redistribute it and/or modify
%%    it under the terms of the GNU General Public License version 3 as
%%    published by the Free Software Foundation.
%%
%%    RCFS is distributed in the hope that it will be useful,
%%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
%%    GNU General Public License for more details.
%%
%%    You should have received a copy of the GNU General Public License
%%    along with RCFS. If not, see <http://www.gnu.org/licenses/>.

function LQT_CP

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.nbData = 100; %Number of datapoints
param.nbPoints = 3; %Number of viapoints
param.nbVarPos = 2; %Dimension of position data (here: x1,x2)
param.nbDeriv = 2; %Number of static and dynamic features (nbDeriv=2 for [x,dx] and u=ddx)
param.nbVar = param.nbVarPos * param.nbDeriv; %Dimension of state vector
param.dt = 1E-2; %Time step duration
param.nbFct = 12; %Number of basis functions
param.basisName = 'RBF'; %PIECEWISE, RBF, BERNSTEIN, FOURIER
param.r = 1E-8; %Control cost in LQR 

%Task setting (viapoints passing)
Mu = [rand(param.nbVarPos, param.nbPoints) - 0.5; zeros(param.nbVar-param.nbVarPos, param.nbPoints)]; %Viapoints
%Q = speye(param.nbVar * param.nbPoints) * 1E0; %Precision matrix (for full state)
Q = kron(eye(param.nbPoints), diag([ones(1,param.nbVarPos) * 1E0, zeros(1,param.nbVar-param.nbVarPos)])); %Precision matrix (for position only)
R = speye((param.nbData-1)*param.nbVarPos) * param.r; %Standard control weight matrix (at trajectory level)

%Time occurrence of viapoints
tl = linspace(1, param.nbData, param.nbPoints+1);
tl = round(tl(2:end));
idx = (tl - 1) * param.nbVar + [1:param.nbVar]';

%Dynamical System settings (discrete version)
A1d = zeros(param.nbDeriv);
for i=0:param.nbDeriv-1
	A1d = A1d + diag(ones(param.nbDeriv-i,1),i) * param.dt^i * 1/factorial(i); %Discrete 1D
end
B1d = zeros(param.nbDeriv,1); 
for i=1:param.nbDeriv
	B1d(param.nbDeriv-i+1) = param.dt^i * 1/factorial(i); %Discrete 1D
end
A = kron(A1d, speye(param.nbVarPos)); %Discrete nD
B = kron(B1d, speye(param.nbVarPos)); %Discrete nD

%Build Sx and Su transfer matrices
Sx0 = kron(ones(param.nbData,1), speye(param.nbVar));
Su0 = sparse(param.nbVar*param.nbData, param.nbVarPos*(param.nbData-1));
M = B;
for n=2:param.nbData
	id1 = (n-1)*param.nbVar+1:param.nbData*param.nbVar;
	Sx0(id1,:) = Sx0(id1,:) * A;
	id1 = (n-1)*param.nbVar+1:n*param.nbVar; 
	id2 = 1:(n-1)*param.nbVarPos;
	Su0(id1,id2) = M;
	M = [A*M(:,1:param.nbVarPos), M]; 
end
Su = Su0(idx,:);
Sx = Sx0(idx,:);

%Build basis functions
if isequal(param.basisName,'PIECEWISE')
	phi = buildPhiPiecewise(param.nbData-1, param.nbFct);
elseif isequal(param.basisName,'RBF')
	phi = buildPhiRBF(param.nbData-1, param.nbFct);
elseif isequal(param.basisName,'BERNSTEIN')
	phi = buildPhiBernstein(param.nbData-1, param.nbFct);
elseif isequal(param.basisName,'FOURIER')
	phi = buildPhiFourier(param.nbData-1, param.nbFct);
end

%Application of basis functions to multidimensional control commands
Psi = kron(phi, eye(param.nbVarPos)); 


%% Linear quadratic tracking (batch formulation)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
x0 = zeros(param.nbVar,1); %Initial state
w = (Psi' * Su' * Q * Su * Psi + Psi' * R * Psi) \ Psi' * Su' * Q * (Mu(:) - Sx * x0); %Superposition weights
u = Psi * w; %Control commands
rx = reshape(Sx0*x0+Su0*u, param.nbVar, param.nbData); %Reproduced trajectory
ru = reshape(u, param.nbVarPos, param.nbData-1);    

%Post-processing for plotting
if isequal(param.basisName,'FOURIER')
	rx = real(rx);
	ru = real(ru);
	phi = real(phi); %for plotting
end


%% Plot 2D
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
h(1) = figure('position',[10 10 800 800]); hold on; axis off;
plot(rx(1,:), rx(2,:), 'k-','linewidth',2);
plot(rx(1,1), rx(2,1), 'k.','markersize',30);
plot(Mu(1,:), Mu(2,:), 'r.','markersize',30);
axis equal; 


%% Timeline plot
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clrmap = lines(param.nbFct);
h(2) = figure('position',[950 10 800 800]); 
%States plot
for j=1:param.nbVarPos
	subplot(2*param.nbVarPos+1,1,j); hold on;
	plot(rx(j,:), 'k-','linewidth',2);
	plot(rx(j,1), 'k.','markersize',30);
	plot(tl, Mu(j,:), 'r.','markersize',30);
	ylabel(['x_' num2str(j)], 'fontsize',26);
end
%Commands plot
for j=1:param.nbVarPos
	subplot(2*param.nbVarPos+1,1,param.nbVarPos+j); hold on;
	plot(ru(j,:), 'k-','linewidth',2);
	ylabel(['u_' num2str(j)], 'fontsize',26);
end
%Basis functions plot
subplot(2*param.nbVarPos+1,1,2*param.nbVarPos+1); hold on; 
for i=1:param.nbFct
	plot(phi(:,i), '-','linewidth',3,'color',clrmap(i,:));
end
xlabel('t','fontsize',26); 
ylabel('\phi_k','fontsize',26);

waitfor(h)
end


%Building piecewise constant basis functions
function phi = buildPhiPiecewise(nbData, nbFct) 
	phi = kron(eye(nbFct), ones(ceil(nbData/nbFct),1));
	phi = phi(1:nbData,:);
end

%Building radial basis functions (RBFs)
function phi = buildPhiRBF(nbData, nbFct) 
	t = linspace(0, 1, nbData);
	tMu = linspace(t(1)-1/(nbFct-3), t(end)+1/(nbFct-3), nbFct); %Repartition of centers to limit border effects
	sigma = 1 / (nbFct-2); %Standard deviation
	phi = exp(-(t' - tMu).^2 / sigma^2);
	
	%Optional rescaling
	%phi = phi ./ repmat(sum(phi,2), 1, nbFct); 
end

%Building Bernstein basis functions
function phi = buildPhiBernstein(nbData, nbFct)
	t = linspace(0, 1, nbData);
	phi = zeros(nbData, nbFct);
	for i=1:nbFct
		phi(:,i) = factorial(nbFct-1) ./ (factorial(i-1) .* factorial(nbFct-i)) .* (1-t).^(nbFct-i) .* t.^(i-1);
	end
end

%Building Fourier basis functions
function phi = buildPhiFourier(nbData, nbFct)
	t = linspace(0, 1, nbData);
	
	%Computation for general signals (incl. complex numbers)
	d = ceil((nbFct-1)/2);
	k = -d:d;
	phi = exp(t' * k * 2 * pi * 1i); 
	%phi = cos(t' * k * 2 * pi); %Alternative computation for real signal
		
%	%Alternative computation for real and even signal
%	k = 0:nbFct-1;
%	phi = cos(t' * k * 2 * pi);
%	%phi(:,2:end) = phi(:,2:end) * 2;
%	%invPhi = cos(k' * t * 2 * pi) / nbData;
%	%invPsi = kron(invPhi, eye(param.nbVar));
end
